boards/BMS/Src/Io/adc_driver.c

```diff
@@ -0,0 +1,178 @@
+#define LOW_RES_TS_CURRENT_OFFSET_SCALED (OUTPUT_2_OFFSET_V * OUTPUT_2_SENSITIVITY)
+#define LOW_RES_TS_CURRENT_SCALE (SINGLE_END_RAW_ADC_TO_V * OUTPUT_2_DIV * OUTPUT_2_SENSITIVITY)
+
+// A macro that can be used to get the ADC result
+#define GET_CONVERTED_ADC_VALUE(raw, scale, offset) ((raw) * (scale) + (offset))
+
+typedef enum
+{
+    ADC_INSTANCE_1 = 0U,
+    ADC_INSTANCE_2,
+} AdcInstance_E;
+
+// In STM32 terminology, each ADC pin corresponds to an ADC channel (See:
+// ADCEx_channels). If there are multiple ADC channels being measured, the ADC
+// measures them based on "rank" (See: ADCEx_regular_rank). The rank starts
+// counting at 1, and the lower the rank, the higher the measurement priority.
+//
+// For example, suppose we are measuring ADC channel 2, 4, and 7, which have
+// rank 3, 1, and 2 respectively. The ADC will measure the channel 4, then
+// channel 7, and finally channel 2. This order is important because it
+// determines the order in which the DMA writes data to our raw_adc2_values
+// buffer.
+//
+// The following enum is used to index into raw_adc2_values, which means it must
+// be ordered in ascending ranks. If we were writing an enum for the earlier
+// example, it would look like:
+//
+// enum
+// {
+//     CHANNEL_4, // Rank 1
+//     CHANNEL_7, // Rank 2
+//     CHANNEL_2, // Rank 3
+//     NUM_ADC_CHANNELS,
+// };
+typedef enum
+{
+    RANK_ADC1_CHANNEL_3 = 0U,
+    NUM_ADC1_CHANNELS,
+    RANK_ADC2_CHANNEL_1 = 0U,
+    RANK_ADC2_CHANNEL_3,
+    NUM_ADC2_CHANNELS,
+} AdcRank_E;
+
+typedef struct
+{
+    AdcInstance_E adc_instance; // Specify which ADC the channel corresponds to
+    uint8_t       rank;         // Specifies the rank of the ADC channel
+    float         scale;        // The scale used to convert the raw ADC result to a "real" result
+    float         offset;       // An optional offset field used while converting a raw ADC result to a "real" result
+} AdcConfig_S;
+
+extern ADC_HandleTypeDef hadc1;
+extern ADC_HandleTypeDef hadc2;
+extern TIM_HandleTypeDef htim3;
+
+// Buffers to store raw ADC values
+static uint16_t raw_adc1_buffer[NUM_ADC1_CHANNELS] = { 0U };
+static uint16_t raw_adc2_buffer[NUM_ADC2_CHANNELS] = { 0U };
+
+// Buffers to store converted ADC values
+static float    adc1_voltages[NUM_ADC1_CHANNELS]   = { 0U };
+static float    adc2_voltages[NUM_ADC2_CHANNELS]   = { 0U };
+
+static AdcConfig_S adc_cfg[NUM_ADC_CHANNELS] =
+{
+    // ADC instance 1 configuration
+    [TS_VOLTAGE] =
+    {
+        .adc_instance = ADC_INSTANCE_1,
+        .rank = RANK_ADC1_CHANNEL_3,
+        .scale = TS_VOLTAGE_SCALE,
+        .offset = TS_VOLTAGE_OFFSET_SCALED,
+    },
+
+    // ADC instance 2 configuration
+    [HIGH_RES_TS_CURRENT_SENSE] =
+    {
+        .adc_instance = ADC_INSTANCE_2,
+        .rank = RANK_ADC2_CHANNEL_1,
+        .scale = HIGH_RES_TS_CURRENT_SCALE,
+        .offset = HIGH_RES_TS_CURRENT_OFFSET_SCALED,
```

Comment:
Note that the value of the offsets prepared during compile time should be in the units of the "real" value, not voltage or a digital ADC value